ispcrt/CMakeLists.txt

```diff
@@ -52,158 +53,179 @@ if (ISPCRT_BUILD_GPU)
-    find_package(OpenMP REQUIRED)
-    if (OpenMP_FOUND)
-      target_link_libraries(ispcrt_tasking INTERFACE OpenMP::OpenMP_CXX)
-      target_compile_definitions(ispcrt_tasking INTERFACE ISPC_USE_OMP)
-    endif()
-  elseif (ISPCRT_BUILD_TASK_MODEL STREQUAL "TBB")
-      find_package(TBB REQUIRED COMPONENTS tbb)
-      if (TBB_FOUND)
-        # TBB package does not produce any info when found, so print it here.
-        message(STATUS "Found TBB: ${TBB_VERSION} at ${TBB_DIR}")
-        target_link_libraries(ispcrt_tasking INTERFACE TBB::tbb)
-        target_compile_definitions(ispcrt_tasking INTERFACE ISPC_USE_TBB_PARALLEL_FOR)
-      endif()
-  else()
-    find_package(Threads REQUIRED)
-    if (Threads_FOUND)
-      # Use default compile definitions from ispc_tasking.cpp
-      target_link_libraries(ispcrt_tasking INTERFACE Threads::Threads)
-    endif()
-  endif()
+# Security options
+if (MSVC)
+  # Stack canaries
+  target_compile_options(${PROJECT_NAME} PRIVATE /GS)
+  # Control flow guard
+  target_link_options(${PROJECT_NAME} PRIVATE /GUARD:CF)
+elseif (APPLE)
+else()
+  # Enable stack protector
+  # NOT to assume that null pointer deference does not exist
+  # Assume that signed overflow always wraps
+  target_compile_options(${PROJECT_NAME} PRIVATE -fstack-protector-strong -fno-delete-null-pointer-checks -fwrapv)
 endif()
 
-## Define a macro to build both static and shared versions of ispcrt
+## Install targets + exports ##
 
-macro(build_ispcrt SHARED_OR_STATIC TARGET_NAME)
-  ## Build library ##
+set_target_properties(${PROJECT_NAME}
+    PROPERTIES VERSION ${PROJECT_VERSION} SOVERSION ${PROJECT_VERSION_MAJOR})
 
-  add_library(${TARGET_NAME} ${SHARED_OR_STATIC}
-    $<$<BOOL:${ISPCRT_BUILD_TASKING}>:ispc_tasking.cpp>
+install(TARGETS ${PROJECT_NAME}
+  EXPORT ${PROJECT_NAME}_Exports
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    NAMELINK_SKIP
+  # on Windows put the dlls into bin
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+  # ... and the import lib into the devel package
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
 
-    ispcrt.cpp
-    ${CMAKE_CURRENT_LIST_DIR}/../common/version.rc
-    $<$<BOOL:${ISPCRT_BUILD_CPU}>:detail/cpu/CPUDevice.cpp>
-    $<$<BOOL:${ISPCRT_BUILD_GPU}>:detail/gpu/GPUDevice.cpp>
-  )
-  set_target_properties(${TARGET_NAME} PROPERTIES
-    CXX_STANDARD 17
-    CXX_STANDARD_REQUIRED ON
-  )
+install(EXPORT ${PROJECT_NAME}_Exports
+  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}-${PROJECT_VERSION}
+  NAMESPACE ${PROJECT_NAME}::
+)
 
-  if (ISPCRT_BUILD_GPU)
-    target_compile_definitions(${TARGET_NAME} PRIVATE ISPCRT_BUILD_GPU)
-  endif()
+install(TARGETS ${PROJECT_NAME}
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    NAMELINK_ONLY
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
 
-  if (ISPCRT_BUILD_CPU)
-    target_compile_definitions(${TARGET_NAME} PRIVATE ISPCRT_BUILD_CPU)
-  endif()
+# Static library target.
```

Comment:
Could you combine duplicated logic for static and shared library to the macro as it was before?

Reply:
Shared/static target logic become more different than before, e.g. list of files, include directories, an extra define, so I don't thinks the macro would be OK. I may set the common part in one place looping through ${PROJECT_NAME} ${PROJECT_NAME}_static.

Reply:
I'm asking about combing common part only. Setting of C++ standard, security options, some compile definitions and part of install logic are 100% identical.

Reply:
I have fixed that.